Implemented Implicited Regularized Gaussians

jeverink opened the pull request:
closes #342

amal-ghamdi commented:
Thanks @jeverink for adding the PR :)

nabriis commented:
Hello Jasper, 

As we talked about duringt the hackathon we need to extend the implicit prior a bit to work with the Gibbs using this underlying Gaussian structure. To get a first version working, please add this to the bottom of the `_implicit_regularized_gaussian.py` file

```python
    # This is a getter only attribute for the underlying Gaussian
    # It also ensures that the name of the underlying Gaussian
    # matches the name of the implicit regularized Gaussian
    @property
    def gaussian(self):
        if self._name is not None:
            self._gaussian._name = self._name
        return self._gaussian

    # --- Defer behavior of the underlying Gaussian --- #
    @property
    def geometry(self):
        return self.gaussian.geometry
    
    @geometry.setter
    def geometry(self, value):
        self._gaussian.geometry = value

    def get_conditioning_variables(self):
        return self.gaussian.get_conditioning_variables()
    
    def get_mutable_variables(self):
        return self.gaussian.get_mutable_variables()
    
    @property
    def mean(self):
        return self.gaussian.mean
    
    @mean.setter
    def mean(self, value):
        self.gaussian.mean = value

    @property
    def cov(self):
        return self.gaussian.cov
    
    @cov.setter
    def cov(self, value):
        self.gaussian.cov = value

    @property
    def prec(self):
        return self.gaussian.prec
    
    @prec.setter
    def prec(self, value):
        self.gaussian.prec = value

    @property
    def sqrtprec(self):
        return self.gaussian.sqrtprec

    @sqrtprec.setter
    def sqrtprec(self, value):
        self.gaussian.sqrtprec = value

    @property
    def sqrtcov(self):
        return self.gaussian.sqrtcov

    @sqrtcov.setter
    def sqrtcov(self, value):
        self.gaussian.sqrtcov = value  
    
    # Overwrite the condition method such that the underlying Gaussian is conditioned in general, except when conditioning on self.name
    # which means we convert Distribution to Likelihood or EvaluatedDensity.
    def _condition(self, *args, **kwargs):

        # When conditioning, we always do it on a copy to avoid unintentional side effects
        new_density = self._make_copy()

        # Check if self.name is in the provided keyword arguments.
        # If so, pop it and store its value.
        value = kwargs.pop(self.name, None)

        new_density._gaussian = self.gaussian._condition(*args, **kwargs)

        # If self.name was provided, we convert to a likelihood or evaluated density
        if value is not None:
            new_density = new_density.to_likelihood(value)

        return new_density
 ```

jeverink commented:
Thanks for helping out @nabriis , the nonnegative Gibbs sampler now works.

I would prefer if mean, cov, prec sqrtcov and sqrtprec to be inaccessible from ImplicitRegularizedGaussian, as these are properties of the underlying Gaussian distribution and not the ImplicitRegularizedGaussian, but it seems that that will be necessary for now.

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,157 @@
+
+
+class ImplicitRegularizedGaussian(Distribution):
+    def __init__(self, gaussian, proximal = None, projector = None, constraint = None, regularization = None, **kwargs):
```
Could you add a docstring that explains this distribution and its parameters? Look e.g. at Gaussian or GMRF docstring.

nabriis reviewed (changes requested):
Thank you very much @jeverink for adding this cool and rather complete set of updates to support implicit regularized Gaussians all the way from the Sampler class to adding it to BayesianProblem 🎉 

Since we are on vacation this week I have just a few quick comments for now in case you are on the office and can continue on this.

1) A few places are missing a docstring. It would be very helpful for other users of these features to have some general information on how it works. This information also gets compiled into the online docs. Can you make sure to add docstrings to the classes that are missing it (I tried to mark afew)

2) Can you add some unit tests that test the features you are adding? In particular I was thinking to add the following (do not worry if you can't add all right now, we can work on it together also):
    - A simple regression test for the FISTA solver like those in https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_solver.py
    - A few unit tests on the implicit regularized linear RTO sampler. You can check https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_sampler.py and maybe https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_problem.py for some inspiration.
    - Try adding some unit tests for the implicit regularized Gaussian distribution. See e.g. https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_distribution.py
    - Add at least one case to the Bayesian Inversion regression test here: https://github.com/CUQI-DTU/CUQIpy/blob/main/tests/test_bayesian_inversion.py for your example.
3) Finally there are 3 unit tests that are failing at this moment. Do not worry about those for now. They are related to the input parameters to distributions, which are generally assumed to be scalars or ndarrays with any other parameters always having a default. I will think a bit on how that works for the Implicit distribution. It may be that we would change input from `ImplicitRegularizedGaussian(Gaussian(mean=0, cov=1), proximal="nonneg")` to something like `ImplicitRegularizedGaussian(mean=0, cov=1, proximal="nonneg")`

nabriis reviewed (changes requested):
Hello again @jeverink. Hope you have had the chance to also hold some vacation.

I have discussed with the CUQIpy team and this PR very close to being merged. It is looking really great and you have incorporated so well into the existing CUQIpy codebase.

There are however two minor changes we would like:

1) We have created a repository for the CUQIpy-User-Showcases where we host jupyter notebooks from user projects / examples. The Implicit_priors notebook is a great fit here and we would like to put it there. If you have happy with the current version of the notebook I can simply copy it over to the repository (or invite you if you prefer to do it yourself). Let me know :) Afterwards the notebook should be removed from this PR.
2) We have agreed that the ImplicitRegularizedGaussian would best with the rest of the CUQIpy distributions if it directly takes the Gaussian parameter arguments as initialization input. This means the __init__ method should be updated to take `mean=None, cov=None, prec=None, sqrtcov=None, sqrtprec=None` rather than a `gaussian` distribution. I believe you can still initialize the Gaussian in the __init__ method using these parameters and then leave the remainder of the getters and setter as we already have them.

Let me know how time commitments look for you this next week. I can also take over and do this minor changes to the PR if you have busy with other tasks and hand over the final review to Jakob.

Thanks again :)

jeverink commented:
Hi @nabriis ,

Thanks for the feedback, I'll look into making the minor changes in the coming week.

jeverink commented:
@nabriis ,

One issue I have with the proposed modification of initializing ImplicitRegularizedGaussian is that the ImplicitRegularizedGaussian can take either a Gaussian distribution or a GMRF distribution. Would I then need to create two separate distributions: ImplicitRegularizedGaussian and ImplicitRegularizedGMRF?

nabriis commented:
> @nabriis ,
> 
> One issue I have with the proposed modification of initializing ImplicitRegularizedGaussian is that the ImplicitRegularizedGaussian can take either a Gaussian distribution or a GMRF distribution. Would I then need to create two separate distributions: ImplicitRegularizedGaussian and ImplicitRegularizedGMRF?

Hi @jeverink. Unfortunately this is mostly a GMRF vs Gaussian issue. I think the best solution for now is to have two implicit regularized diats for Gaussian and GMRF, yes. We wil work on merging the underlying code for GMRF and Gaussian in the future and can then revisit this implicit regularized distribution idea more generally

jeverink commented:
Hi @nabriis ,
I've updated the initialization of the ImplicitRegularizedGaussian distribution and added the ImplicitRegularizedGMRF to facilitate both the Gaussian and GMRF initialization. The implementation of ImplicitRegularizedGMRF is mostly a copy-paste job from ImplicitRegularizedGaussian, so please do check if you know a less clumsy solution.

I have also updated the example notebook to work with both distributions. So you could move that notebook over to the user showcase repository.

Let me know if you need any more changes.

nabriis reviewed (approved):
Hey @jeverink. As discussed I went in and moved the notebook to the user-showcase repository. Furthermore, I did a few minor tweaks here and there to the code. I would say its ready now for @jakobsj  to review!

amal-ghamdi commented on cuqi/solver/_solver.py:
```diff
@@ -565,3 +565,102 @@ class PDHG(object):
+    stepsize : The stepsize of the gradient step.
+    abstol : The numerical tolerance for convergence checks.
+    adapative : Whether to use FISTA or ISTA.
+    """  
```
An example code block will be helpful here. For example (this is mainly similar to what is in the test)
```
    Example
    -----------
    .. code-block:: python

        From cuqi.solver import FISTA,  ProximalL1
        import scipy as sp

        m, n = 10, 5
        A = rng.standard_normal((m, n))
        b = rng.standard_normal(m)
        stepsize = 0.99/(sp.linalg.interpolative.estimate_spectral_norm(A)**2)
        x0 = np.zeros(n)
        fista = FISTA(A, b, x0, proximal = ProximalL1, stepsize = stepsize, maxit = 100, abstol=1e-12, adaptive = True)
        sol, _ = fista.solve()


```

nabriis replied:
Done

amal-ghamdi reviewed (commented):
Thank you @jeverink for adding this rich feature! This is definitely and extensive and very useful addition to CUQIpy!. And thank you @nabriis for your all your input and helping out.

I went through the PR and it looks great, well commented and documented with thorough tests! I do not have any fundamental comments. I mede small suggestions here and there for your consideration, mostly style, readability, and coding conventions.

nabriis commented on cuqi/sampler/_conjugate.py:
```diff
@@ -9,28 +9,46 @@ class Conjugate: # TODO: Subclass from Sampler once updated
-        dist = Gamma(shape=m/2+alpha,rate=.5*np.linalg.norm(L@(Ax-b))**2+beta)
-
-        return dist.sample()
+        if isinstance(self.target.likelihood.distribution, (Gaussian, GMRF)):
+            # Extract variables
+            b = self.target.likelihood.data                                 #mu
+            m = len(b)                                                      #n
+            Ax = self.target.likelihood.distribution.mean                   #x_i
+            L = self.target.likelihood.distribution(np.array([1])).sqrtprec #L
+            alpha = self.target.prior.shape                                 #alpha
+            beta = self.target.prior.rate                                   #beta
+    
+            # Create Gamma distribution and sample
+            dist = Gamma(shape=m/2+alpha,rate=.5*np.linalg.norm(L@(Ax-b))**2+beta)
+    
+            return dist.sample()
+        elif isinstance(self.target.likelihood.distribution, (ImplicitRegularizedGaussian, ImplicitRegularizedGMRF)):
+            # Extract variables
+            b = self.target.likelihood.data                                          #mu
+            m = len(b)                                                               #n
+            Ax = self.target.likelihood.distribution.gaussian.mean                   #x_i
+            L = self.target.likelihood.distribution(np.array([1])).gaussian.sqrtprec #L
+            alpha = self.target.prior.shape                                          #alpha
+            beta = self.target.prior.rate                                            #beta
+    
+            # Create Gamma distribution and sample
+            dist = Gamma(shape=np.count_nonzero(b)/2+alpha,rate=.5*np.linalg.norm(L@(Ax-b))**2+beta)
+            return dist.sample()
```
@jeverink, regarding the `np.count_nonzero(b)`. Is it fundamentally a different conjugate update compared to `len(b)`? Would adding constraints change how this update was carried out? Do you perhaps have a reference I can look up?

jeverink replied:
Yes, it is fundamentally different. The sampler guarantees that some samples will have zeros, hence len(b) is greater than np.count_nonzero(b) most of the times. See our paper [Bayesian inference with projected densities](https://epubs.siam.org/doi/abs/10.1137/22M150695X) for details.

nabriis replied:
Great. Updated

amal-ghamdi replied:
Thanks for the update 👍

nabriis commented:
Thanks @amal-ghamdi for the great comments. I will look at the comments and update accordingly tonight. There a few things I need to check with @jeverink, and I will "@" him in those.

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        # Preset information, for use in Gibbs
+        self._preset = None
+        
+        if proximal is not None:
+            self._proximal = proximal
+        elif projector is not None:
+            self._proximal = lambda z, gamma: projector(z)
+        elif (isinstance(constraint, str) and constraint.lower().replace("-","") in ["nonnegativity", "nonnegative", "nn"]):
+            self._proximal = lambda z, gamma: ProjectNonnegative(z)
+            self._preset = "nonnegativity"
+        elif (isinstance(constraint, str) and constraint.lower() in ["box"]):
+            self._proximal = lambda z, gamma: ProjectBox(z)
+            self._preset = "box" # Not supported in Gibbs
+        elif (isinstance(regularization, str) and regularization.lower() in ["l1"]):
+            self._proximal = ProximalL1
+            self._preset = "l1"
+        else:
+            raise ValueError("Regularization not supported")
```
My understanding is that a projection can also be viewed as a proximal. E.g. it is a subset of all possible proximal operators. On the other hand the "regularization" is more like a string input, which returns a proximal. So the proximal name is correct.

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        if proximal is not None:
+            if not callable(proximal):
+                raise ValueError("Proximal needs to be callable.")
+            if len(get_non_default_args(proximal)) != 2:
```
@jeverink is this a use case or would the sampling always pass in both x and scale? The reason we kept it like this was to support this case:
```python
def f(x, scale, extra_par=10):
    # do computation
    # retrun value
```

jeverink replied:
The solvers will always pass both x and scale.

nabriis replied:
Then the code should be as it is.

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+    projector : callable f(x) or None
+        Euclidean projection onto the constraint
+
+    constraint : string or None, Default: "nonnegative"
+        Preset constraints, including "nonnegativity" and "box". Required for use in Gibbs.
```
Agreed, will simplify to only one type of argument

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        Euclidean proximal operator of the regularization function R
+
+    projector : callable f(x) or None
+        Euclidean projection onto the constraint
```
@jeverink, can you write a short sentence to put in the doc?

amal-ghamdi replied:
👍

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        If a scalar or 1d-array the value is assumed to be the inverse standard deviation of each component of the Gaussian.
+
+    proximal : callable f(x, scale) or None
+        Euclidean proximal operator of the regularization function R
```
@jeverink, can you write a short sentence to put in the doc?

jeverink replied:
I have added a small description for the proximal and projector.

amal-ghamdi replied:
Thanks for the update :)

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+
+    sqrtprec : scalar or 1d-array or 2d-array (sparse matrix is supported)
+        Square root of precision matrix of Gaussian. Defined as matrix R, where R.T@R = prec.
+        If a scalar or 1d-array the value is assumed to be the inverse standard deviation of each component of the Gaussian.
```
Good point!

amal-ghamdi replied:
thanks for the update :)

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+    in the form of a proximal operator or a projector. Alternatively, preset constraints and regularization
+    can be used.
+
+    Only one of proximal, projector, constraint or regularization can be provided. If none of them are provided,
```
We always have a default set of parameters other than scaler and vector. Considering it is a regularized gaussian, it would be a bit strange to have the default be no regularization? Then I would rather actually force the user to specify a type of regularization. Thoughtts @amal-ghamdi ?

amal-ghamdi replied:
It could be better to require the user to specify regularization so they are aware of the impact. But this is just my suggestion, you can alternatively leave it as it is if you think that is better.

nabriis replied:
I will let @jakobsj  weight in. This would be the first distribution in CUQIpy that does not a default argument for non-vector type parameters that can also be None. E.g. we can do

```python
x = Gaussian(0, 1)
y = GMRF(0, 1)
```
etc.

jakobsj replied:
Just saw this, please make sure this is resolved ie no default is used. Perhaps just the docstring needs updating in line with the gaussian version.

nabriis replied:
Just docstring yes

nabriis commented on cuqi/distribution/__init__.py:
```diff
@@ -5,6 +5,7 @@
 from ._gamma import Gamma
 from ._gaussian import Gaussian, JointGaussianSqrtPrec
 from ._gmrf import GMRF
+from ._implicit_regularized_gaussian import ImplicitRegularizedGaussian, ImplicitRegularizedGMRF
```
We agreed it was an implicit prior. I would consider leaving it in distribution module for now. On a similar topic though, how about the name? @amal-ghamdi. Maybe it should be shorter? Cant really come up with any good shorthands though.

nabriis replied:
Yes. Lets leave this for now.

jeverink commented on cuqi/solver/_solver.py:
```diff
@@ -565,3 +565,102 @@ class PDHG(object):
+class FISTA(object):
+    """Fast Iterative Shrinkage-Thresholding Algorithm for regularized least squares problems.
+
+    Minimize ||Ax-b||^2 + f(x).
```
I have added a reference

amal-ghamdi replied:
👍

nabriis commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        if projector is not None:
+            if not callable(projector):
+                raise ValueError("Projector needs to be callable.")
+            if len(get_non_default_args(projector)) != 1:
```
See above

nabriis commented on cuqi/problem/_problem.py:
```diff
@@ -351,6 +351,10 @@ def sample_posterior(self, Ns, Nb=None, callback=None) -> cuqi.samples.Samples:
             return self._samplepCN(Ns, Nb, callback)
+        
+        # For Regularized Gaussians with linear models we use Regularized LinearRTO
+        elif self._check_posterior(self, (ImplicitRegularizedGaussian, ImplicitRegularizedGMRF), Gaussian):
```
Yes. Updated. Used slightly different syntax.

nabriis commented on cuqi/sampler/_rto.py:
```diff
@@ -165,3 +167,151 @@ def _sample(self, N, Nb):
+        Maximum number of iterations of the inner FISTA solver. *Optional*.
+        
+    stepsize : string or float
+        If stepsize is a string and equals either "auto", "automatic" or "spectral_norm", then the stepsize is automatically estimated based on the spectral norm.
```
Good idea

nabriis commented on cuqi/sampler/_rto.py:
```diff
@@ -165,3 +167,151 @@ def _sample(self, N, Nb):
     def _sample_adapt(self, N, Nb):
         return self._sample(N,Nb)
+
+class RegularizedLinearRTO(Sampler):
```
#320

amal-ghamdi replied:
👍

nabriis commented on cuqi/solver/_solver.py:
```diff
@@ -565,3 +565,102 @@ class PDHG(object):
+        return x, k
+    
+    
+def ProjectNonnegative(x):
+    """(Euclidean) projection onto the nonnegative orthant.
+    
+    Parameters
+    ----------
+    x : array_like.
+    """  
+    return np.maximum(x, 0)
+
+def ProjectBox(x, lower = None, upper = None):
+    """(Euclidean) projection onto a box.
+    
+    Parameters
+    ----------
+    x : array_like.
+    lower : array_like. Lower bound of box. Zero if None.
+    upper : array_like. Upper bound of box. One if None.
+    """  
+    if lower is None:
+        lower = np.zeros_like(x)
+    
+    if upper is None:
+        upper = np.ones_like(x)
+    
+    return np.minimum(np.maximum(x, lower), upper)
+
+def ProximalL1(x, gamma):
+    """(Euclidean) proximal operator of the \|x\|_1 norm.
+    Also known as the shrinkage or soft thresholding operator.
+    
+    Parameters
+    ----------
+    x : array_like.
+    gamma : scale parameter.
+    """
+    return np.multiply(np.sign(x), np.maximum(np.abs(x)-gamma, 0))
```
Ok sounds good to me. In a way these are solvers for simple optimization problems :)

amal-ghamdi replied:
👍

nabriis commented:
Thanks @amal-ghamdi. I have updated according to your comments and suggestions

amal-ghamdi commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,273 @@
+        
+    def __init__(self, mean=None, cov=None, prec=None, sqrtcov=None, sqrtprec=None, proximal = None, projector = None, constraint = None, regularization = None, **kwargs):
+        
+        # We init the underlying Gaussian first for geometry and dimensionality handling
+        self._gaussian = Gaussian(mean=mean, cov=cov, prec=prec, sqrtcov=sqrtcov, sqrtprec=sqrtprec, **kwargs)
+        
+        # Init from abstract distribution class
+        super().__init__(**kwargs)
```
Got it!

amal-ghamdi commented on cuqi/solver/_solver.py:
```diff
@@ -565,3 +565,102 @@ class PDHG(object):
+        if not callable(A):
+            self.explicitA = True
+        else:
+            self.explicitA = False
```
thanks for the update :)

amal-ghamdi commented on cuqi/distribution/_implicit_regularized_gaussian.py:
```diff
@@ -0,0 +1,278 @@
+    sqrtprec
+        See :class:`~cuqi.distribution.Gaussian` for details.
+
+    proximal : callable f(x, scale) or None
+        Euclidean proximal operator of the regularization function g, that is, a solver for the optimization problem
+        min_z 0.5||x-z||_2^2+scale*g(x).
```
```suggestion
    proximal : callable f(x, scale) or None
        Euclidean proximal operator of the regularization function f, that is, a solver for the optimization problem
        min_z 0.5||x-z||_2^2+scale*f(x).

```

jeverink replied:
Thanks for the suggestion. I chose not to use f in the optimization problem as it is a different function from the f in ` proximal : callable f(x, scale)`

amal-ghamdi replied:
Got it! Thanks for the clarification. I editied my comment (just wanted to emphasize where f comes into play)

amal-ghamdi reviewed (approved):
Many thanks for the updates @jeverink and @nabriis. This PR is really a great addition to CUQpy 👏👏👏.

I approved (there are three unresolved comments, feel free to take a look and resolve as you see fit).

jakobsj reviewed (changes requested):
Excellent work, very comprehensive new set of features, and I really like how it allows users to provide their own projectors/proximals while also supplying existing ones using string input. I have read through the notebook (that is now in a showcase PR) and the code here and I have number of (initial?) comments and questions:

1. My understanding is that the Implicit Regularized Gaussian (and GMRF) should not be considered distributions, but implicit priors.  For example, and as the notebook demonstrates, one cannot sample from it, and also not get logpdf etc. As such, it seems these implicit priors do not belong as part of the distribution module (even if they inherit from distribution) but rather in a separate module, such "implicitprior" to distinguish from actual distributions. Aligned with this, from a software perspective, a common convention for subclassing is that child classes should still admit the parent class methods, but here .sample, .logpdf etc are not.
2. I think - in the spirit of clarity and being explicit following our recent discussions - it is better to force the user to specify a regularizer/constraint than letting nonnegativity be the default. I see the point that we currently provide defaults for all our distributions, but if you agree with the point above, then this is not a distribution and as such may not follow that (informal) convention.
3. Accordingly, in the code, comments and notebook, a fixed terminology should be used, as it is now sometimes referred to as Implicit Regularized Gaussian  distribution or prior or just Gaussian. In particular, I think "distribution" should be avoided. 
4. I think the proximals should be put somewhere dedicated, not directly in cuqi.solver, as they are not solvers. Could be something like cuqi.solver.proximal or rename as cuqi.opti.proximal, and could consider moving e.g. FISTA, PDHG etc. to cuqi.opti.algo or similar, since "solver" is a bit vague, could mean equation solver or other solver, and I think all are optimization algorithms. 
5. Is the "box" constraint implemented specifically as the interval [0,1] or how does one provide the upper and lower bounds for general box constraints? I haven't heard of the convention that box means [0,1], but consider a general interval to be specified by the user. In the simplest case for scalar values, as [value1, value2] or e.g. lower=value1, upper=value2, more generally value1 and value2 could be vectors (and contain plus/minus inf).
6. Docstring specifies that only one of constraint, projector, proximal can be passed. Please state what happens if the user does attempt this - is an error thrown or does one of them take precedence (my pref would be throwing an error).
7. In the notebook, in the first cell with ProximalL1, why use "6" in "ProximalL1(z, 6*s)", the text does not explain this number? Is it just another parameter? Also the 0.1 just before, I couldn't see that explained - maybe state the relevant formula with this number mentioned to clarify?
8. Do we want to allow multiple different string values "nonnegative", "NN", "nonnegativity"? I think in general one is better, because otherwise, when to stop "nonneg", "NNEG", "nonnegative_values", ...  Question is which one of course... My preference would probably be "nonnegativity", as this is a noun, like "box" stating what type of constraint, ie, you wouldn't say "nonnegative constraint", since the constraint *itself* is not nonnegative.

I'll leave it at this for now - again, impressive work, very much looking forward to getting these features merged in, to provide the features directly available in the PR, but also the general approach it opens up for, incl PnP priors etc.

jeverink commented:
Thanks for all the feedback @jakobsj ,

Regarding (2) and (6):
Defaulting to any specific constraints or regularization seems confusing. I now throw an error in case zero or more than one of the required parameters are provided.

Regarding (5), the box constraints:
I have now added optional keyword arguments to allow the user to specify lower and upper bounds, i.e, 
`x = ImplicitRegularizedGMRF(np.zeros(n), 200, constraint="box", lower_bound = 0.1, upper_bound = 0.9)`.
If None, these will default to 0 and 1, respectively. (I probably implemented this very clumsily, so @nabriis , could you look at it?)

Regarding (7):
I can update the notebook, but I forgot to accept the invitation to the user_showcase repo you send some time ago. Can I be reinvited to this repo?

nabriis commented:
Hey @jeverink. I am planing to resolve the merge conflicts and incorporate the remainder of Jakobs comments including refactoring the classes into their own module. I wanted to give you a heads up, as I see you are still working and pushing the the branch. Let me know if you have any final updates or changes. I will push my changes sometime tomorrow or so.

jeverink commented:
Thanks for the head-up @nabriis.
My final commits were, besides addressing some of Jakobs comments, some minor fixes and a small requested feature.
I have not planned any further updates yet.

nabriis commented:
Hello @jakobsj  and @jeverink. I now refactored and fixed conflicts with main. I also updated the regression tests that needed to be updated after the bug fix to FISTA. As for jakobs point 4) 

"I think the proximals should be put somewhere dedicated, not directly in cuqi.solver, as they are not solvers. Could be something like cuqi.solver.proximal or rename as cuqi.opti.proximal, and could consider moving e.g. FISTA, PDHG etc. to cuqi.opti.algo or similar, since "solver" is a bit vague, could mean equation solver or other solver, and I think all are optimization algorithms."

I would like to postpone fixing this and creating an issue to fix it. It would be a rather substantial change that would be best handled in a new PR. I would also like to avoid creating a new temporary module (e.g. cuqi.proximals) which we may delete after refactoring cuqi.solver.

jeverink commented:
Thanks for the refactoring and updates @nabriis ,
I updated a minor error message, but besides that I am happy with the current version.
I also updated the notebook in the User-Showcase to work with the current version

jakobsj commented on cuqi/implicitprior/_regularizedGMRF.py:
```diff
@@ -0,0 +1,71 @@
+class RegularizedGMRF(RegularizedGaussian):
+    """ Implicit Regularized GMRF (Gaussian Markov Random Field). 
+
+    Defines a GMRF distribution with implicit regularization defining a so-called implicit prior.
```
```suggestion
    Defines a so-called implicit prior based on a GMRF distribution with implicit regularization.
```

jakobsj commented on cuqi/implicitprior/_regularizedGMRF.py:
```diff
@@ -0,0 +1,71 @@
+                    "upper_bound" : kwargs.pop("upper_bound", None),
+                    "strength" : kwargs.pop("strength", None)}
+            
+            # Underlying explicit Gaussian
```
```suggestion
            # Underlying explicit GMRF
```

jakobsj commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -0,0 +1,236 @@
+class RegularizedGaussian(Distribution):
+    """ Implicit Regularized Gaussian.
+
+    Defines a Gaussian distribution with implicit regularization creating a so-called implicit prior.
```
```suggestion
    Defines a so-called implicit prior based on a Gaussian distribution with implicit regularization.
```

jakobsj commented on cuqi/problem/_problem.py:
```diff
@@ -347,6 +348,10 @@ def sample_posterior(self, Ns, Nb=None, callback=None) -> cuqi.samples.Samples:
         elif self._check_posterior(self, (Gaussian, GMRF), Gaussian):
             return self._samplepCN(Ns, Nb, callback)
+        
+        # For Regularized Gaussians with linear models we use Regularized LinearRTO
```
```suggestion
        # For Regularized Gaussians with linear models we use RegularizedLinearRTO
```

jakobsj reviewed (changes requested):
Thank you @nabriis for the updates.  I have gone through and it looks very good. Added a couple of minor suggestions just to consistently avoid referring to the implicit priors as distributions.  I agree the proximals can be moved later, please create the issue.

nabriis commented:
@jakobsj ready for final review

jakobsj reviewed (approved):
Looks good, thanks!
